Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,239 @@
+example (hfa : FnLb f a) (hgb : FnLb g b) : FnLb (fun x ↦ f x + g x) (a + b) :=
+  sorry
+
+-- 2. Now try proving it without using any tactic!
+
+example (hfa : FnLb f a) (hgb : FnLb g b) : FnLb (fun x ↦ f x + g x) (a + b) :=
+  sorry
+
+-- 3. Here you may need the theorem `mul_nonneg`
+
+#check mul_nonneg
+
+example (nnf : FnLb f 0) (nng : FnLb g 0) : FnLb (fun x ↦ f x * g x) 0 :=
+  sorry
+
+
+
+
+
+/-!
+  # Existential quantifier
+-/
+
+-- To write `∃` type
+
+/-!
+  To prove an exists statement, we need to provide 2 things. An object of
+  the correct type and a proof that it satisfies the desired property.
+  The first step is done with `use`
+-/
+
+ example : ∃ x : ℝ, 2 * x = 5 := by
+  use 5/2 -- Notice how the goal has changed
+  norm_num -- This is a useful tactic that does arithmetic computations.
+
+-- We can construct a proof term explicitly using the "anonymous constructor" notation
+
+example : ∃ x : ℝ, 2 * x = 5 :=
+  ⟨5/2 , by norm_num⟩ -- Lean matches the given data with the goal
+
+-- To use an exists statement, we need the command `rcases` and the anonymus constructor
+
+def FnHasUb (f : ℝ → ℝ) :=
+  ∃ a, FnUb f a -- f has an upper bound
+
+example (ubf : FnHasUb f) (ubg : FnHasUb g) : FnHasUb fun x ↦ f x + g x := by
+  rcases ubf with ⟨a, ubfa⟩
+  -- Adds new variable a (the upper bound) with property ubfa (a is the upper bound)
+  rcases ubg with ⟨b, ubgb⟩
+  use a + b
+  apply fnUb_add ubfa ubgb
+
+  -- Intro and rcases can be combined into one command `rintro`
+
+example : FnHasUb f → FnHasUb g → FnHasUb fun x ↦ f x + g x := by
+  rintro ⟨a, ubfa⟩ ⟨b, ubgb⟩
+  exact ⟨a + b, fnUb_add ubfa ubgb⟩
+
+
+/-!
+# Exercises
+-/
+
+open Function
+
+-- 4. The definition of surjective function involves a universal and an exist quantifier.
+-- `Surjective (f : ℝ → ℝ) := ∀ y, ∃ x, f x = y`
+-- Prove the following using both `intro` and `rcases`
+
+example {c : ℝ} : Surjective fun x ↦ x + c := by
+  sorry
+
+-- 5. Now the exist statement is given as hypothesis. Prove the goal using `rcases`
+
+example (surjg : Surjective g) (surjf : Surjective f) : Surjective fun x ↦ g (f x) := by
+  sorry
+
+
+
+/-!
+  # Negation
+-/
+
+-- To write `¬` type
+-- Lean interprets a negation `¬a` as `a → False`, so tactics for implications work
+
+
+example (h : Q) : ¬¬Q := by
+  intro notQ -- This creates the hypothesis `¬Q` and changes the goal to `False`
+  exact notQ h
+  -- We treat `¬Q` as an implication, so we can feed it a proof for `Q` to get a proof of `False`
+
+
+/-!
+  Lean comes with many tactics that helps to deal with negation:
+  · `by_contra h` proof by contradiction: allows us to prove `Q` by assuming
+    `h=¬Q` and deriving a contradiction
+  ⬝ `push_neg` push negations inside quantifiers, changing `¬∀` into `∃¬`, etc...
+  ⬝ `contrapose` transforms a goal `A → B` to `¬B → ¬A`
+  ⬝ `contrapose!` does `contrapose`, and then applies `push_neg` automatically
+  ⬝ `exfalso` applies the theorem `False → Q`, changing the goal into `False`
+-/
```

Comment:
```suggestion
/-!
  Lean comes with many tactics that helps to deal with negation:
  · `by_contra h` proof by contradiction: allows us to prove `Q` by assuming
    `h = ¬ Q` and deriving a contradiction
  ⬝ `push_neg` push negations inside quantifiers, changing `¬ ∀` into `∃ ¬`, etc...
  ⬝ `contrapose` transforms a goal `A → B` to `¬ B → ¬ A`
  ⬝ `contrapose!` does `contrapose`, and then applies `push_neg` automatically
  ⬝ `exfalso` applies the theorem `False → Q`, changing the goal into `False`
-/
```

Reply:
I was following the stylistic convention of MIL that doesn't put a space between the negation and the proposition